src/rtlib.jl

```diff
@@ -66,7 +66,8 @@ end
 function emit_function!(mod, config::CompilerConfig, f, method)
     tt = Base.to_tuple_type(method.types)
     source = generic_methodinstance(f, tt)
-    new_mod, meta = codegen(:llvm, CompilerJob(source, config); toplevel=false)
+    new_mod, meta = codegen(:llvm, CompilerJob(source, config);
+        optimize=false, libraries=false, validate=false)
```

Comment:
That shouldn't be necessary; the idea was that `toplevel` drives the default setting of `optimize`, `libraries` and `validate`: https://github.com/JuliaGPU/GPUCompiler.jl/blob/ae993ba5592ba143724de4d96027f42505ae8270/src/driver.jl#L82-L84

Reply:
I think with `optimize=false, libraries=false, validate=false` we still get `toplevel=true` and `cleanup=true` (while with `toplevel=false` all of them are `false`).

When `toplevel=true` we execute `finish_ir!` on a module, but for AMDGPU we treat it like a late library linking:
https://github.com/JuliaGPU/AMDGPU.jl/blob/1c08e7ac9f243d81fd19900c1554f918e71030bc/src/compiler/codegen.jl#L43

Because at library linking stage (among other things) we [specify wavefront size](https://github.com/JuliaGPU/AMDGPU.jl/blob/1c08e7ac9f243d81fd19900c1554f918e71030bc/src/compiler/device_libs.jl#L62) which affects codegen.
So when `toplevel=false` this does not happen at all and by default it emits `wave64`.

However, proper `wave64` on `wave32` native hardware is enabled in LLVM 17: https://github.com/llvm/llvm-project/commit/07c59204878045b58ea1fe976ac2aac58d2efbb6

So here we get:
```llvm
ERROR: LLVM error: Cannot select: 0x5310bd0: i64 = SETCC 0x84bf590, Constant:i32<0>, setne:ch
  0x84bf590: i32 = zero_extend # D:1 0x5311030
    0x5311030: i1,ch = CopyFromReg # D:1 0x667d070, Register:i1 %36
      0x86d35f0: i1 = Register %36
  0x5310a10: i32 = Constant<0>
In function: __ockl_dm_trim
```

But even if we were on LLVM 17 we still should link those libraries (although they are moving to instead specifying them via target machine or llvm ir attributes).

Reply:
I see. In some sense the runtime library really isn't a top-level compilation job though, so I thought it made sense to make it `toplevel=false`. Why doesn't the call to `finish_ir!` that happens when compiling the subsequent kernel that links against the runtime make the IR compatible?

Reply:
Not sure... maybe it is too late or needs to be done for every module...

Reply:
I'll try to see if linking options libraries can be moved to finish_module! instead

Reply:
There's a specific `link_libraries!` phase too, doesn't that work?

Reply:
Yeah and we use it, but it is not called when `toplevel` is false

Reply:
Ah, right. It really looks like `finish_ir!` is the correct phase to put this in then, I'd maybe rather look why the transformations in there aren't sufficient to legalize the runtime IR that's been linked in? Otherwise there might be duplicate work when `process_module!` is called multiple times.

Reply:
e.g., if the issue is that `finish_ir!` happens _after_ IR optimization, you could put the linking step in `optimize_module!`, which is also only executed once.